File: src/main/java/com/ptaf/action_performer/ActionPerformer.java
package com.ptaf.action_performer;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.MouseButton;
import com.microsoft.playwright.options.WaitForSelectorState;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.file.Paths;

/**
 * ActionPerformer is a utility class for performing a wide variety of actions
 * on Playwright Locators within a Page. It handles clicks, form input, attribute manipulation,
 * element validation, scrolling, screenshots, and much more.
 *
 * It also supports returning values such as text content, attributes, or boolean results
 * when applicable, while performing non-returning actions like click or hover as void-equivalent.
 */
public class ActionPerformer {

    private static final Logger logger = LoggerFactory.getLogger(ActionPerformer.class);

    /**
     * Utility to format text content by placing each word on a new line.
     * Useful for improving text readability in logs or reports.
     */
    private static String formatTextForNewLine(String text) {
        return text.replaceAll(" ", "\n");
    }

    /**
     * Performs the specified action on a target Locator and returns the result if applicable.
     *
     * @param page          Playwright Page instance
     * @param action        The action name (e.g., "click", "gettext", "select")
     * @param targetLocator The element to perform the action on
     * @param value         An optional value for the action (e.g., text to fill, attribute name)
     * @return              A String result if the action yields one (e.g., gettext), otherwise null
     */
    public String performActionWithReturn(Page page, String action, Locator targetLocator, String value) {
        return performAction(page, action, targetLocator, value);
    }

    /**
     * Performs the specified action on a target Locator and returns the result if applicable.
     */
    public String performAction(Page page, String action, Locator targetLocator, String value) {
        try {
            switch (action.toLowerCase()) {
                case "click": targetLocator.click(); return null;
                case "fill": targetLocator.fill(value); return null;
                case "select": targetLocator.selectOption(value); return null;
                case "selectmultiple": targetLocator.selectOption(value.split(",")); return null;
                case "check": targetLocator.check(); return null;
                case "uncheck": targetLocator.uncheck(); return null;
                case "hover": targetLocator.hover(); return null;
                case "type": targetLocator.type(value); return null;
                case "press": targetLocator.press(value); return null;
                case "dblclick": targetLocator.dblclick(); return null;
                case "rightclick": targetLocator.click(new Locator.ClickOptions().setButton(MouseButton.valueOf("right"))); return null;
                case "tap": targetLocator.tap(); return null;
                case "input": targetLocator.evaluate("(element, val) => element.value = val", value); return null;
                case "screenshot": targetLocator.screenshot(new Locator.ScreenshotOptions().setPath(Paths.get(value))); return null;
                case "scroll": targetLocator.evaluate("element => element.scrollIntoView({ behavior: 'smooth', block: 'center' })"); return null;
                case "focus": targetLocator.focus(); return null;
                case "blur": targetLocator.evaluate("element => element.blur()"); return null;
                case "clear": targetLocator.clear(); return null;
                case "drag": Locator target = targetLocator.page().locator(value); targetLocator.dragTo(target); return null;
                case "dragstart": targetLocator.dispatchEvent("dragstart"); return null;
                case "dragend": targetLocator.dispatchEvent("dragend"); return null;
                case "uploadfile":
                case "selectfile": targetLocator.setInputFiles(Paths.get(value)); return null;
                case "file_chooser_for_upload": page.waitForFileChooser(() -> click(targetLocator)); return null;
                case "getattribute": return targetLocator.getAttribute(value);
                case "setattribute": targetLocator.evaluate("(el, val) => el.setAttribute('value', val)", value); return null;
                case "removeattribute": targetLocator.evaluate("(el, attr) => el.removeAttribute(attr)", value); return null;
                case "gettext": return targetLocator.textContent();
                case "get_and_contain_text": String getText = targetLocator.textContent(); assertCondition(getText.contains(getText), "Element does not contain expected text."); return getText;
                case "getvalue": return targetLocator.inputValue();
                case "hasvalue": String currentValue = targetLocator.inputValue(); assertCondition(currentValue.equals(value), "Expected: " + value + ", but found: " + currentValue); return currentValue;
                case "isvisible": boolean isVisible = targetLocator.isVisible(); assertCondition(isVisible, "Element is not visible."); return String.valueOf(isVisible);
                case "isenabled": boolean isEnabled = targetLocator.isEnabled(); assertCondition(isEnabled, "Element is not enabled."); return String.valueOf(isEnabled);
                case "ischecked": boolean isChecked = targetLocator.isChecked(); assertCondition(isChecked, "Element is not checked."); return String.valueOf(isChecked);
                case "isdisabled": boolean isDisabled = targetLocator.isDisabled(); assertCondition(isDisabled, "Element is not disabled."); return String.valueOf(isDisabled);
                case "ishidden": boolean isHidden = targetLocator.isHidden(); assertCondition(isHidden, "Element is not hidden."); return String.valueOf(isHidden);
                case "exists": boolean exists = targetLocator.count() > 0; assertCondition(exists, "Element does not exist."); return String.valueOf(exists);
                case "not_exists": boolean notExists = targetLocator.count() == 0; assertCondition(notExists, "Element exists but should not."); return String.valueOf(notExists);
                case "hastext": String locatorText = targetLocator.textContent(); assertCondition(locatorText.contains(value), "Text mismatch."); return locatorText;
                case "hasclass": boolean hasClass = targetLocator.getAttribute("class").contains(value); assertCondition(hasClass, "Class mismatch."); return String.valueOf(hasClass);
                case "hasequalvalue": String actualValue = targetLocator.inputValue(); assertCondition(actualValue.equals(value), "Value mismatch."); return actualValue;
                case "isempty": String inputValue = targetLocator.inputValue(); assertCondition(inputValue.isEmpty(), "Element is not empty."); return inputValue;
                case "waitforelement": targetLocator.waitFor(); return null;
                case "waitforstate": targetLocator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.valueOf(value.toUpperCase()))); return null;
                case "waitfortext": targetLocator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE)); if (!targetLocator.textContent().contains(value)) throw new AssertionError("Text not found: " + value); return targetLocator.textContent();
                case "waitforvalue": targetLocator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE)); if (!targetLocator.inputValue().equals(value)) throw new AssertionError("Value mismatch."); return targetLocator.inputValue();
                case "evaluate": targetLocator.evaluate(value); return null;
                default: throw new IllegalArgumentException("Unknown action: " + action);
            }
        } catch (Exception e) {
            logger.error("Error while performing action: {} for Target Locator {}", action, targetLocator, e);
            throw new RuntimeException("Action failed: " + action + " for Target Locator: " + targetLocator + " - " + e.getMessage(), e);
        }
    }

    private void click(Locator targetLocator) {
        try {
            targetLocator.click();
        } catch (Exception e) {
            logger.error("Error while clicking on target locator: {}", e.getMessage());
            throw new RuntimeException("Click action failed: " + e.getMessage(), e);
        }
    }

    private void assertCondition(boolean condition, String errorMessage) {
        if (!condition) {
            throw new AssertionError(errorMessage);
        }
    }

    public void waitForLocator(Locator locator) {
        try {
            locator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE).setTimeout(60000.0));
        } catch (Exception e) {
            logger.error("Failed to wait for the element to be displayed", e);
        }
    }
}

File: src/main/ptaf/ui/action_performer.py
"""Playwright locator action dispatch (ported from ActionPerformer.java)."""

from __future__ import annotations

import logging
from pathlib import Path

from playwright.sync_api import Locator, Page

logger = logging.getLogger(__name__)


class ActionPerformer:
    """Perform actions on Playwright locators within a page context."""

    def perform_action_with_return(
        self,
        page: Page,
        action: str,
        target_locator: Locator,
        value: str | None,
    ) -> str | None:
        return self.perform_action(page, action, target_locator, value)

    def perform_action(
        self,
        page: Page,
        action: str,
        target_locator: Locator,
        value: str | None,
    ) -> str | None:
        try:
            action_key = action.lower()
            first = target_locator.first

            match action_key:
                case "click":
                    first.click()
                    return None
                case "fill":
                    first.fill(value or "")
                    return None
                case "select":
                    first.select_option(value or "")
                    return None
                case "selectmultiple":
                    first.select_option((value or "").split(","))
                    return None
                case "check":
                    first.check()
                    return None
                case "uncheck":
                    first.uncheck()
                    return None
                case "hover":
                    first.hover()
                    return None
                case "type":
                    first.type(value or "")
                    return None
                case "press":
                    first.press(value or "")
                    return None
                case "dblclick":
                    first.dblclick()
                    return None
                case "rightclick":
                    first.click(button="right")
                    return None
                case "tap":
                    first.tap()
                    return None
                case "input":
                    first.evaluate("(element, val) => element.value = val", value)
                    return None
                case "screenshot":
                    first.screenshot(path=Path(value or ""))
                    return None
                case "scroll":
                    first.evaluate(
                        "element => element.scrollIntoView({ behavior: 'smooth', block: 'center' })"
                    )
                    return None
                case "focus":
                    first.focus()
                    return None
                case "blur":
                    first.evaluate("element => element.blur()")
                    return None
                case "clear":
                    first.clear()
                    return None
                case "drag":
                    target = target_locator.page.locator(value or "").first
                    first.drag_to(target)
                    return None
                case "dragstart":
                    first.dispatch_event("dragstart")
                    return None
                case "dragend":
                    first.dispatch_event("dragend")
                    return None
                case "uploadfile" | "selectfile":
                    first.set_input_files(Path(value or ""))
                    return None
                case "download":
                    with page.expect_download() as download_info:
                        first.click()
                    download = download_info.value
                    save_path = Path(value or "")
                    download.save_as(save_path)
                    logger.info(
                        "File downloaded successfully and saved to: %s", save_path
                    )
                    return str(download.path())
                case "file_chooser_for_upload":
                    with page.expect_file_chooser():
                        self._click(first)
                    return None
                case "setattribute":
                    first.evaluate("(el, val) => el.setAttribute('value', val)", value)
                    return None
                case "removeattribute":
                    first.evaluate(
                        "(el, attr) => el.removeAttribute(attr)", value
                    )
                    return None
                case "evaluate":
                    first.evaluate(value or "")
                    return None
                case "waitforelement":
                    first.wait_for()
                    return None
                case "waitforstate":
                    first.wait_for(state=(value or "").lower())
                    return None
                case "getattribute":
                    return first.get_attribute(value or "")
                case "gettext":
                    return first.text_content()
                case "get_and_contain_text":
                    get_text = first.text_content()
                    self._assert_condition(
                        get_text is not None and get_text in get_text,
                        "Element does not contain expected text.",
                    )
                    return get_text
                case "getvalue":
                    return first.input_value()
                case "hasvalue":
                    current_value = first.input_value()
                    self._assert_condition(
                        current_value == value,
                        f"Expected: {value}, but found: {current_value}",
                    )
                    return current_value
                case "isvisible":
                    is_visible = first.is_visible()
                    self._assert_condition(
                        is_visible, "Element is not visible."
                    )
                    return str(is_visible)
                case "isenabled":
                    is_enabled = first.is_enabled()
                    self._assert_condition(
                        is_enabled, "Element is not enabled."
                    )
                    return str(is_enabled)
                case "ischecked":
                    is_checked = first.is_checked()
                    self._assert_condition(
                        is_checked, "Element is not checked."
                    )
                    return str(is_checked)
                case "isdisabled":
                    is_disabled = first.is_disabled()
                    self._assert_condition(
                        is_disabled, "Element is not disabled."
                    )
                    return str(is_disabled)
                case "ishidden":
                    is_hidden = first.is_hidden()
                    self._assert_condition(
                        is_hidden, "Element is not hidden."
                    )
                    return str(is_hidden)
                case "hastext":
                    locator_text = first.text_content()
                    self._assert_condition(
                        locator_text is not None and value in locator_text,
                        "Text mismatch.",
                    )
                    return locator_text
                case "hasclass":
                    class_attr = first.get_attribute("class")
                    has_class = class_attr is not None and value in class_attr
                    self._assert_condition(has_class, "Class mismatch.")
                    return str(has_class)
                case "hasequalvalue":
                    actual_value = first.input_value()
                    self._assert_condition(
                        actual_value == value, "Value mismatch."
                    )
                    return actual_value
                case "isempty":
                    input_value = first.input_value()
                    self._assert_condition(
                        input_value == "", "Element is not empty."
                    )
                    return input_value
                case "waitfortext":
                    first.wait_for(state="visible")
                    text_content = first.text_content()
                    if text_content is None or value not in text_content:
                        raise AssertionError(f"Text not found: {value}")
                    return text_content
                case "waitforvalue":
                    first.wait_for(state="visible")
                    input_value = first.input_value()
                    if input_value != value:
                        raise AssertionError("Value mismatch.")
                    return input_value
                case "exists":
                    exists = target_locator.count() > 0
                    self._assert_condition(
                        exists, "Element does not exist."
                    )
                    return str(exists)
                case "not_exists":
                    not_exists = target_locator.count() == 0
                    self._assert_condition(
                        not_exists, "Element exists but should not."
                    )
                    return str(not_exists)
                case _:
                    raise ValueError(f"Unknown action: {action}")

        except Exception as exc:
            logger.error(
                "Error while performing action: %s for Target Locator %s",
                action,
                target_locator,
                exc_info=True,
            )
            raise RuntimeError(
                f"Action failed: {action} for Target Locator: {target_locator} - {exc}"
            ) from exc

    def _click(self, target_locator: Locator) -> None:
        try:
            target_locator.click()
        except Exception as exc:
            logger.error(
                "Error while clicking on target locator: %s", exc, exc_info=True
            )
            raise RuntimeError(f"Click action failed: {exc}") from exc

    @staticmethod
    def _assert_condition(condition: bool, error_message: str) -> None:
        if not condition:
            raise AssertionError(error_message)

    def wait_for_locator(self, locator: Locator) -> None:
        try:
            locator.first.wait_for(state="visible", timeout=60_000)
        except Exception:
            logger.error(
                "Failed to wait for the element to be displayed", exc_info=True
            )
